File: EnvMonitorAndroid/app/src/main/java/com/envmonitor/models/CurrentData.kt
package com.envmonitor.models

/**
 * Mô hình dữ liệu hiện tại từ thiết bị giám sát
 * 
 * @property temperature Dữ liệu nhiệt độ
 * @property humidity Dữ liệu độ ẩm
 * @property pm10 Dữ liệu bụi PM10
 * @property pm25 Dữ liệu bụi PM2.5
 * @property co Dữ liệu khí CO
 * @property noise Dữ liệu tiếng ồn
 * @property aqi Dữ liệu chỉ số chất lượng không khí (AQI)
 * @property deviceStatus Trạng thái thiết bị (online, inactive, offline)
 * @property lastDataTimestamp Thời điểm nhận dữ liệu mới nhất
 */
data class CurrentData(
    val temperature: EnvParameter,
    val humidity: EnvParameter,
    val pm10: EnvParameter,
    val pm25: EnvParameter,
    val co: EnvParameter,
    val noise: EnvParameter,
    val aqi: EnvParameter,
    val deviceStatus: String,
    val lastDataTimestamp: Long
) {
    
    /**
     * Lấy danh sách các thông số dưới dạng các mục để hiển thị trên giao diện
     * 
     * @return Danh sách các thông số có thể hiển thị trên UI
     */
    fun toParameterList(): List<ParameterCardItem> {
        val displayNames = mapOf(
            "temperature" to "Nhiệt độ",
            "humidity" to "Độ ẩm",
            "pm10" to "PM10",
            "pm25" to "PM2.5",
            "co" to "CO",
            "noise" to "Tiếng ồn",
            "aqi" to "AQI"
        )
        
        return ParameterCardItem.fromCurrentData(this, displayNames)
    }
    
    /**
     * Lấy đối tượng trạng thái thiết bị
     * 
     * @return Đối tượng DeviceStatus chứa thông tin trạng thái thiết bị
     */
    fun getDeviceStatusObject(): DeviceStatus {
        return DeviceStatus.fromServerStatus(deviceStatus, lastDataTimestamp)
    }
    
    /**
     * Kiểm tra xem có thông số nào vượt ngưỡng cảnh báo hay không
     * 
     * @return true nếu có ít nhất một thông số vượt ngưỡng cảnh báo
     */
    fun hasAnyWarning(): Boolean {
        return temperature.isWarningOrAbove() || 
               humidity.isWarningOrAbove() || 
               pm10.isWarningOrAbove() || 
               pm25.isWarningOrAbove() || 
               co.isWarningOrAbove() || 
               noise.isWarningOrAbove() ||
               aqi.isWarningOrAbove()
    }
    
    /**
     * Kiểm tra xem có thông số nào vượt ngưỡng nguy hiểm hay không
     * 
     * @return true nếu có ít nhất một thông số vượt ngưỡng nguy hiểm
     */
    fun hasAnyDanger(): Boolean {
        return temperature.isDanger() || 
               humidity.isDanger() || 
               pm10.isDanger() || 
               pm25.isDanger() || 
               co.isDanger() || 
               noise.isDanger() ||
               aqi.isDanger()
    }
}

File: env_monitor_app/lib/models/current_data.dart
import 'parameter_card_item.dart';
import 'device_status.dart';

class CurrentData {
  final Map<String, dynamic> data;
  final String timestamp;
  
  CurrentData({
    required this.data,
    required this.timestamp,
  });
  
  factory CurrentData.fromJson(Map<String, dynamic> json) {
    return CurrentData(
      data: json['data'] as Map<String, dynamic>,
      timestamp: json['timestamp'].toString(),
    );
  }
  
  List<ParameterCardItem> toParameterList() {
    final List<ParameterCardItem> result = [];
    
    // Các thông số có trong data
    if (data.containsKey('temperature')) {
      result.add(ParameterCardItem.fromJson('temperature', data['temperature'] as Map<String, dynamic>));
    }
    
    if (data.containsKey('humidity')) {
      result.add(ParameterCardItem.fromJson('humidity', data['humidity'] as Map<String, dynamic>));
    }
    
    if (data.containsKey('pm25')) {
      result.add(ParameterCardItem.fromJson('pm25', data['pm25'] as Map<String, dynamic>));
    }
    
    if (data.containsKey('pm10')) {
      result.add(ParameterCardItem.fromJson('pm10', data['pm10'] as Map<String, dynamic>));
    }
    
    if (data.containsKey('co')) {
      result.add(ParameterCardItem.fromJson('co', data['co'] as Map<String, dynamic>));
    }
    
    if (data.containsKey('noise')) {
      result.add(ParameterCardItem.fromJson('noise', data['noise'] as Map<String, dynamic>));
    }
    
    if (data.containsKey('aqi')) {
      result.add(ParameterCardItem.fromJson('aqi', data['aqi'] as Map<String, dynamic>));
    }
    
    return result;
  }
  
  DeviceStatus getDeviceStatusObject() {
    return DeviceStatus(
      status: data.containsKey('device_status') ? data['device_status'] as String : 'unknown',
      lastDataTimestamp: data.containsKey('last_data_timestamp') 
          ? (data['last_data_timestamp'] as num).toInt()
          : 0,
    );
  }
  
  bool isDeviceOnline() {
    return data.containsKey('device_status') && data['device_status'] == 'online';
  }
}